### PR TITLE
Improving CH type system with concepts

### PR DESCRIPTION
**Pull request by myrrc**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Not for changelog (changelog entry is not required)

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

1. Remove hardcoded toString/typeToString method for enums with magic_enum contrib.
2. Simplify wrappers readText and writeText.
3. Make use of is_decimal and is_integral concepts.

### CODE REVIEW

File: base/common/EnumReflection.h [hunk 1]
```diff
@@ -0,0 +1,32 @@
+#pragma once
+
+#include <type_traits>
+#include <fmt/format.h>
+#include <../../contrib/magic_enum/include/magic_enum.hpp> //FIXME
+
+// Can't name concept is_enum as is would conflict with type trait
+template <class T> concept type_is_enum = std::is_enum_v<T>;
```
Comment (amosbird) [inline base/common/EnumReflection.h:8]: May be the first PR containing `concept`)

Comment (alexey-milovidov) [reply base/common/EnumReflection.h:8]: Ok if it's supported by (both) our compilers...

Comment (myrrc) [reply base/common/EnumReflection.h:8]: There were issues with gcc and some google contrib, I'll fix them if they arise again.

File: base/common/EnumReflection.h [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+};
+
+namespace DB
+{
+
+/**
+ * Some enums have multiple-word values like FLUSH_DICT. However, we should parse user input like FLUSH DICT.
+ * magic_enum::enum_names returns names with underscore, so we have to replace underscore with space.
+ */
+inline std::string UnderscoreToSpace(std::string_view str)
```
Comment (alexey-milovidov) [inline base/common/EnumReflection.h:28]: UnderscoreToSpace -> underscoreToSpace

File: base/common/EnumReflection.h [hunk 3]
```diff
@@ -0,0 +1,34 @@
+#pragma once
+
+#include <type_traits>
+#include <fmt/format.h>
+#include <../../contrib/magic_enum/include/magic_enum.hpp> //FIXME
```
Comment (alexey-milovidov) [inline base/common/EnumReflection.h:5]: Fix before merge?

File: base/common/EnumReflection.h [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+
+// Can't name concept is_enum as is would conflict with type trait
+template <class T> concept type_is_enum = std::is_enum_v<T>;
+
+// Enable printing enum values as strings via fmt + magic_enum
+template <type_is_enum T>
```
Comment (alexey-milovidov) [inline base/common/EnumReflection.h:11]: Cool!

File: contrib/CMakeLists.txt [hunk 5]
```diff
@@ -32,6 +32,8 @@ endif()
 
 set_property(DIRECTORY PROPERTY EXCLUDE_FROM_ALL 1)
 
+include_directories("${CMAKE_SOURCE_DIR}/contrib/magic_enum/include")
```
Comment (alexey-milovidov) [inline contrib/CMakeLists.txt:35]: I'm not sure if it works here, maybe it's only for descendant CMakeFiles?

File: src/Common/ErrorCodes.cpp [hunk 6]
```diff
@@ -13,6 +13,8 @@
   * - system.errors table
   */
 
+// FIXME Unused UNKNOWN_VOLUME_TYPE
```
Comment (alexey-milovidov) [inline src/Common/ErrorCodes.cpp:16]: Just remove it.

File: contrib/CMakeLists.txt (hunk 5 above)
Comment (myrrc) [reply contrib/CMakeLists.txt:35]: I guess it doesn't, but I failed the fight with CMake to be able to just `#include<magic_enum.hpp>` in EnumReflection.h, it won't work.

Comment (alexey-milovidov) [reply contrib/CMakeLists.txt:35]: Should be doable... most likely we have some other header only libs in contrib.

Comment (alexey-milovidov) [reply contrib/CMakeLists.txt:35]: Maybe put into `target_include_directories` for `base`?

File: base/common/find_symbols.h [hunk 7]
```diff
@@ -66,19 +60,13 @@ inline __m128i mm_is_in(__m128i bytes)
...
-    if constexpr (positive)
-        return x;
-    else
-        return !x;
-}
+template <bool do_not_negate>
+constexpr bool maybe_negate(bool x) { return x && do_not_negate; }
 
-template <bool positive>
+template <bool do_not_negate>
```
Comment (alexey-milovidov) [inline base/common/find_symbols.h:66]: How is it better than "positive"?

File: src/Functions/URL/extractURLParameter.cpp [hunk 8]
```diff
@@ -32,7 +32,7 @@ struct ExtractURLParameterImpl
             /// Find query string or fragment identifier.
             /// Note that we support parameters in fragment identifier in the same way as in query string.
 
-            const char * const query_string_begin = find_first_symbols<'?', '#'>(str, end);
+            const char * const query_string_begin = find_first_symbols<"?#">(str, end);
```
Comment (alexey-milovidov) [inline src/Functions/URL/extractURLParameter.cpp:35]: I don't see any advantage in this change.

File: base/common/StringLiteral.h [hunk 9]
```diff
@@ -0,0 +1,16 @@
+#pragma once
+
+#include <algorithm>
+
+/// Meta-struct that can be used in template parameters:
+/// template <StringLiteral sl> void foo();
+/// foo<"hello">()
+template<size_t N>
+struct StringLiteral {
+    constexpr StringLiteral(const char (&str)[N]) {
```
Comment (alexey-milovidov) [inline base/common/StringLiteral.h:10]: Style.

File: src/Functions/URL/extractURLParameter.cpp (hunk 8 above)
Comment (alexey-milovidov) [reply src/Functions/URL/extractURLParameter.cpp:35]: But ok, let's do it.

File: contrib/CMakeLists.txt (hunk 5 above)
Comment (azat) [reply contrib/CMakeLists.txt:35]: You can add an INTERFACE library, and link it to the dbms (or clickhouse_common_io), you can take a look at how pcg_random is added - https://github.com/ClickHouse/ClickHouse/blob/3e65fa1a10bc9d489aec536c5b270856b767ac25/base/pcg-random/CMakeLists.txt#L1-L2

Comment (myrrc) [reply contrib/CMakeLists.txt]: It worked, thanks!

**Review by alexey-milovidov — APPROVED** [review]

**Pull request merged**